cli/internal/oauth/manager/manager.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/docker/docker/registry"
+
+	"github.com/pkg/browser"
+)
+
+// OAuthManager is the manager responsible for handling authentication
+// flows with the oauth tenant.
+type OAuthManager struct {
+	store       credentials.Store
+	tenant      string
+	audience    string
+	clientID    string
+	api         api.OAuthAPI
+	openBrowser func(string) error
+}
+
+// OAuthManagerOptions are the options used for New to create a new auth manager.
+type OAuthManagerOptions struct {
+	Store       credentials.Store
+	Audience    string
+	ClientID    string
+	Scopes      []string
+	Tenant      string
+	DeviceName  string
+	OpenBrowser func(string) error
+}
+
+func New(options OAuthManagerOptions) *OAuthManager {
+	scopes := []string{"openid", "offline_access"}
+	if len(options.Scopes) > 0 {
+		scopes = options.Scopes
+	}
+
+	openBrowser := browser.OpenURL
+	if options.OpenBrowser != nil {
+		openBrowser = options.OpenBrowser
+	}
+
+	return &OAuthManager{
+		clientID: options.ClientID,
+		audience: options.Audience,
+		tenant:   options.Tenant,
+		store:    options.Store,
+		api: api.API{
+			TenantURL: "https://" + options.Tenant,
+			ClientID:  options.ClientID,
+			Scopes:    scopes,
+		},
+		openBrowser: openBrowser,
+	}
+}
+
+// LoginDevice launches the device authentication flow with the tenant,
+// printing instructions to the provided writer and attempting to open the
+// browser for the user to authenticate.
+// After the user completes the browser login, LoginDevice uses the retrieved
+// tokens to create a Hub PAT which is returned to the caller.
+// The retrieved tokens are stored in the credentials store (under a separate
+// key), and the refresh token is concatenated with the client ID.
+func (m *OAuthManager) LoginDevice(ctx context.Context, w io.Writer) (*types.AuthConfig, error) {
+	state, err := m.api.GetDeviceCode(ctx, m.audience)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get device code: %w", err)
+	}
+
+	if state.UserCode == "" {
+		return nil, errors.New("no user code returned")
+	}
+
+	_, _ = fmt.Fprintln(w, "\n\033[1mUSING WEB BASED LOGIN\033[0m")
+	_, _ = fmt.Fprintln(w, "To sign in with credentials on the command line, use 'docker login -u <username>'")
+	_, _ = fmt.Fprintf(w, "\nYour one-time device confirmation code is: \033[1m%s\033[0m\n", state.UserCode)
+	_, _ = fmt.Fprintf(w, "\033[1mPress ENTER\033[0m to open your browser or submit your device code here: \033[4m%s\033[0m\n", strings.Split(state.VerificationURI, "?")[0])
+
+	tokenResChan := make(chan api.TokenResponse)
+	waitForTokenErrChan := make(chan error)
+	go func() {
+		tokenRes, err := m.api.WaitForDeviceToken(ctx, state)
+		if err != nil {
+			waitForTokenErrChan <- err
+			return
+		}
+		tokenResChan <- tokenRes
+	}()
+
+	go func() {
+		reader := bufio.NewReader(os.Stdin)
+		reader.ReadString('\n')
+		_ = m.openBrowser(state.VerificationURI)
+	}()
+
+	_, _ = fmt.Fprint(w, "\nWaiting for authentication in the browser…\n")
```

Comment:
Is the single-character triple dot intentional?
Perhaps using a regular 3-character dot would be safer, we also seem to use it in other place:
https://github.com/docker/cli/blob/3b2e2e54ea28f227e61697dc734ee5541bd5af23/cli/command/registry/login.go#L121

```suggestion
	_, _ = fmt.Fprint(w, "\nWaiting for authentication in the browser...\n")
```

Reply:
Right, I wondered about that. I had that initially, and then changed it per @dvdksn's suggestion

Reply:
It is unicode iirc – I think we use some unicode characters in other places, but if we think it's safer to have a regular ellipsis I'm totally fine with that.

Reply:
I'm also fine with a single-character! Just wanted to make sure that it was intentional and not overzealous IDE 😅

Reply:
I _think_ we're already using it in some other places, so (probably?) fine